doc/sphinx/language/gallina-specification-language.rst

```diff
@@ -1045,6 +1045,51 @@ co-inductive definitions are also allowed.
    object of type :g:`(EqSt s1 s2)`. We will see how to introduce infinite
    objects in Section :ref:`cofixpoint`.
 
+Caveat
+++++++
+
+The ability to define co-inductive types by constructors, hereafter called
+*positive co-inductive types*, is known to break subject reduction. It is a
+fundamental issue related to the duality of computational behavior between
+inductive and co-inductive types.
+
+Since the introduction of primitive records in Coq 8.5, an alternative
+presentation is available, called *negative co-inductive types*. This consists
+in defining a co-inductive type as a primitive record type through its
+projections. Such a technique is akin to the *co-pattern* style that can be
+found in e.g. Agda, and preserves subject reduction.
+
+The above example can be rewritten in the following way.
+
+.. coqtop:: all
+
+   Set Primitive Projections.
+   CoInductive Stream : Set := Seq { hd : nat; tl : Stream }.
+   CoInductive EqSt (s1 s2: Stream) : Prop := eqst {
+     eqst_hd : hd s1 = hd s2;
+     eqst_tl : EqSt (tl s1) (tl s2);
+   }.
+
+Some properties that hold over positive streams are lost when going to the
+negative presentation, typically when they imply equality over streams.
```

Comment:
imply or involve?

Reply:
Both are true!

Reply:
@ppedrot: In the absence of answers to previous questions, allow me to ask again: wouldn't it be more correct to say that it is not the move to positive to negative which makes that some properties implying/involving equality are lost but only the amount of η-equality that we allow in each case.

Otherwise said, we could restrict η-equality for the positive formulation or we could add more η-equality for the negative formulation, and this is what matters for the "more equalities over streams".

My understanding of the reason why people tends to consider that negative streams solve 
the extensionality of equality issue leading to a loss of subject reduction is that η is clearly disjoint from the logical rules in the negative case, while it comes implicitly in the dependent form of `match` in the positive case. Couldn't a view be adopted which separates concerns and which questions the relevance of necessarily hard-wiring some form of extensionality in dependent `match`?

Reply:
From what I understand, for a decidable, SR preserving positive treatment of coinductives we'd need to switch to Baclofen type theory or something similar which restricts eta-equalities on the positive coinductives/avoids the implicit eta-equality which is part of dependent elimination. I think we all agree on that no? I suspect having such a variant is part of @ppedrot 's research project as well :) Maybe we could add something to that effect to the reference manual explanation?

Reply:
> SR preserving positive treatment of coinductives we'd need to switch to Baclofen type theory or something similar which restricts eta-equalities on the positive coinductives/avoids the implicit eta-equality which is part of dependent elimination.

That's my understanding, but this is still a rough one. So yes, it boils down to the problem of eta-rules, but all of this is mixed together in the current presentation of type theory.

> Couldn't a view be adopted which separates concerns and which questions the relevance of necessarily hard-wiring some form of extensionality in dependent match?

As explained by @mattam82, that's just BTT. But practical usability of that theory is not clear yet.